### PR TITLE
Improve communication info publisher

### PR DESCRIPTION
**Pull request by fsimonis**
## Main changes of this PR

This PR extends read and write of connection files to be more strict and robust.

The most common issues are now actively checked and result in error messages.
Messages guide the user to either fix or report problems.

## Motivation and additional information

Problems with connection files are tricky to debug as the writers and readers don't actively check conditions.

## Author's checklist

* [ ] I added a changelog file with `make changelog` if there are user-observable changes since the last release.
* [x] I ran `make format` to ensure everything is formatted correctly.
* [x] I sticked to C++14 features.
* [x] I sticked to CMake version 3.16.3.
* [ ] I squashed / am about to squash all commits that should be seen as one.

## Reviewers' checklist

* [ ] Does the changelog entry make sense? Is it formatted correctly?
* [x] Do you understand the code changes?
* [x] Do you agree with the verbosity and the wording?
* [ ] @KyleDavisSA Can you try this on SuperMUC and see if these errors and warnings get triggered by the synchronization delay of filesystem system?

### CODE REVIEW

**Review by uekerman — COMMENTED** [review]
Error messages look good. Only doubt I had was whether users know what you mean with "connection file". Maybe a intro sentence like "There was a problem establishing the m2n communication between participants X and Y." could help?

I guess you have tested to trigger some of them locally? (e.g. with leftover files from a previous run)

**Comment by KyleDavisSA** [comment]
I managed to test this with a few ranks over 2 nodes on superMUC-NG. No problems found. Next I will try something that uses many ranks over more nodes to really test the issue.

**Comment by fsimonis** [comment]
> Only doubt I had was whether users know what you mean with "connection file". Maybe a intro sentence like "There was a problem establishing the m2n communication between participants X and Y." could help?

This is tricky, as this contains information across multiple layers. The connection info publisher isn't aware of the communication, which isn't aware of the participant names.

**Comment by uekerman** [comment]
> This is tricky, as this contains information across multiple layers. The connection info publisher isn't aware of the communication, which isn't aware of the participant names.

OK, but we could still do a generic: "There was a problem establishing a communication channel." or similar.

File: src/com/ConnectionInfoPublisher.cpp [hunk 1]
```diff
@@ -49,56 +50,103 @@ std::string ConnectionInfoPublisher::getLocalDirectory() const
...
-  fs::create_directories(tmp.parent_path());
+  auto path = getFilename();
+  auto tmp  = bfs::path(path + "~");
+
   {
-    std::ofstream ofs(tmp.string(), std::ofstream::out);
-    ofs << info << "\n";
-    ofs << "Acceptor: " << acceptorName << ", Requester: " << requesterName << ", Tag: " << tag << ", Rank: " << rank << "\n";
+    auto message = "Unable to establish connection as a {}connection file already exists at \"{}\". "
+                   "This is likely a leftover of a previous crash during communication build-up. "
```
Comment (uekerman) [inline src/com/ConnectionInfoPublisher.cpp:118]:
```suggestion
                   "This is likely a leftover of a previous crash or stop during communication build-up. "
```

**Review by uekerman — APPROVED** [review]
Did some tests on my laptop. Error messages look good.

File: src/com/ConnectionInfoPublisher.cpp [hunk 2]
```diff
@@ -49,56 +50,103 @@ std::string ConnectionInfoPublisher::getLocalDirectory() const
...
+  }
+  PRECICE_DEBUG("Deleting connection file \"{}\"", path);
   try {
-    fs::path p(getFilename());
-    PRECICE_DEBUG("Deleting connection file {}", p.string());
-    fs::remove(p);
-  } catch (const fs::filesystem_error &e) {
-    PRECICE_WARN("Unable to delete connection file due to error: {}", e.what());
+    bfs::remove(path);
+    /// @TODO Check if we need to sleep to let the underlying filesystem synchronize
```
Comment (DavidSCN) [inline src/com/ConnectionInfoPublisher.cpp:98]: Do you want to check this before merging? How do you want to check this?

**Comment by KyleDavisSA** [comment]
Still struggling to run ASTE on SNG. However the other test cases all ran without issues.

**Review by KyleDavisSA — APPROVED** [review]
Earlier SNG tests ran without issue.

File: src/com/ConnectionInfoPublisher.cpp (hunk 2 above)
Comment (fsimonis) [reply src/com/ConnectionInfoPublisher.cpp:98]: We need to run a test case with a distributed filesystem running on separated nodes preferably with a slow inter-node connection.

**Comment by fsimonis** [comment]
@DavidSCN Do you have a large parallel ASTE setup that you can quickly run on a distributed system?

I really want to avoid this breaking for HPC users :wink:

**Comment by DavidSCN** [comment]
> @DavidSCN Do you have a large parallel ASTE setup that you can quickly run on a distributed system?

Lacking a large cluster. I could run it on our institute machine on 16 x 6 nodes?!

**Comment by fsimonis** [comment]
@DavidSCN Sounds like a perfect fit!

**Review by DavidSCN — APPROVED** [review]
I ran now the test cases using ASTE on our cluster with 42  cores  (6 x 7 nodes) per participant. One testcase was the our most refined reference paper setup (turbine) with a NN mapping, which ran three times in a loop and one was the most coarse reference paper setup with a NN mapping, which ran five times in a loop. Everything worked fine in `Release` and `Debug` mode using the changes proposed here.

**Comment by DavidSCN** [comment]
Could you rebase before merging? Merging will fail due to the fmt upgrade (boost::filesystem::path).

File: src/com/ConnectionInfoPublisher.cpp (hunk 2 above)
Comment (fsimonis) [reply src/com/ConnectionInfoPublisher.cpp:98]: For the record https://github.com/precice/precice/pull/1290#pullrequestreview-1060192506

**Comment by DavidSCN** [comment]
Only a changelog missing, then we are ready to go.

**Pull request merged**